native/dorgtr.go

```diff
@@ -0,0 +1,92 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import "github.com/gonum/blas"
+
+// Dorgtr generates a real orthogonal matrix Q which is defines as the product
+// of n-1 elementary reflectors of order N as returned by Dsytrd.
+//
+// If uplo == blas.Upper
+//  Q = H[n-1] * ... * H[1] * H[0]
+// If uplo == blas.Lower
+//  L = H[0] * ... * H[n-1]
+//
+// Where H[i] is constructed from the elementary reflectors as computed by Dsytrd.
+// Please see the documentation for Dsytrd for more information.
+//
+// tau must have length at least n-1, and Dorgtr will panic otherwise.
+//
+// Work is temporary storage, and lwork specifies the usable memory length. At minimum,
+// lwork >= n-1, and Dorgtr will panic otherwise. The amount of blocking is
+// limited by the usable length.
+// If lwork == -1, instead of computing Dorgtr the optimal work length is stored
+// into work[0].
+func (impl Implementation) Dorgtr(uplo blas.Uplo, n int, a []float64, lda int, tau, work []float64, lwork int) {
+	checkMatrix(n, n, a, lda)
+	if len(tau) < n-1 {
+		panic(badTau)
+	}
+	upper := uplo == blas.Upper
+	var nb int
+	if upper {
+		nb = impl.Ilaenv(1, "DORGQL", " ", n-1, n-1, n-1, -1)
+	} else {
+		nb = impl.Ilaenv(1, "DORGQR", " ", n-1, n-1, n-1, -1)
+		nb = 4
+	}
+	lworkopt := max(1, n-1) * nb
+	work[0] = float64(lworkopt)
+	if lwork == -1 {
+		return
+	}
+	if len(work) < lwork {
```

Comment:
Swap this if with the next one? Let's first check that `lwork` has a good value and then check that work has a good length.